specs/protocol/abi.md

```diff
@@ -6,92 +6,333 @@ This document describes and specifies the ABI (Application Binary Interface) of
+              "name": "<field1_name>",
+              "type": "<field1_type>",
+              "components": ...
+           },
+           {
+              "name": "<field2_name>",
+              "type": "<field2_type>",
+              "components": ...
+           },
+           ...
+       ]
     }
-]
+    ```
+
+    `<field1_type>`, `<field2_type>`, ... are formatted according to the rules of this section. The array `components` for each field is recursively formatted according to the rules of this section.
+
+- `enum`:
+
+    ```json
+    {
+       "name": "<var_name>",
+       "type": "enum <decl_name>",
+       "components": [
+           {
+              "name": "<variant1_name>",
+              "type": "<variant1_type>",
+              "components": ...
+           },
+           {
+              "name": "<variant2_name>",
+              "type": "<variant2_type>",
+              "components": ...
+           },
+           ...
+       ]
+    }
+    ```
+
+    `<variant1_type>`, `<variant2_type>`, ... are formatted according to the rules of this section. The `components` entry for each variant is recursively formatted according to the rules of this section.
+
+- `str[<n>]`:
+
+    ```json
+    {
+       "name": "<var_name>",
+       "type": "str[<n>]",
+       "components": null
+    } 
+    ```
+
+    `<n>` is the size of the string.
+
+- `array`:
+
+    ```json
+    {
+       "name": "<var_name>",
+       "type": "[<element_type>; <n>]",
+       "components": [
+          {
+             "name": "__array_element",
+             "type": "<element_type>",
+             "components": ...
+          }
+       ]
+    },
+    ```
+
+    `<element_type>` is formatted according to the rules of this section. `<n>` is the size of the array. The `components` entry for `__array_element` is recursively formatted according to the rules of this section.
+
+- `tuple`:
+
+    ```json
+    {
+       "name": "<var_name>",
+       "type": "(<field1_type>, <field2_type>, ...)",
+       "components": [
+           {
+              "name": "__tuple_element",
+              "type": "<field1_type>",
+              "components": ...
+           },
+           {
+              "name": "__tuple_element",
+              "type": "<field2_type>",
+              "components": ...
+           },
+           ...
+       ]
+    }
+    ```
+
+    `<field1_type>`, `<field2_type>`, ... are formatted according to the rules of this section. The array `components` for each field is recursively formatted according to the rules of this section.
+
+> **Note**: Because outputs don't have names, the field `"name"` should be set to `""` for each entry of the array `outputs`.
+
+#### Complex example for JSON ABI format
+
+Given the following ABI declaration:
+
+```rust
+enum MyEnum {
+    Foo: u64,
+    Bar: bool,
+}
+struct MyStruct {
+    bim: u8,
+    bam: MyEnum,
+}
+
+abi MyContract {
+    fn complex_function(arg1: MyStruct, arg2: [b256; 4], arg3: (str[5], bool)) -> str[6];
```

Comment:
Can you make this example be an array of structs? It's not actually clear how that would look like.

Reply:
Done.